### PR TITLE
Quiz creation side panels improvements

### LINKED ISSUE #12840: Coach Quiz Side Panels need refactoring
## Overview

In Quiz Creation there is a SectionSidePanel component which wraps a `SidePanelModal` around the `<router-view />` into which the "side panel components" render. This results in clobbering our ability to handle each route individually and complicates routing by putting the onus for some routing decisions on that SectionSidePanel component.

#### Description and outcomes

We should instead simply leverage `VueRouter` and it's guards and keep the logic of "where do I go and why" up to the individual "side panel components" (ie, ResourceSelection, ReplaceQuestions, SectionEditor, etc).

By removing `SectionSidePanel` each of those components then can render _as_ a `SidePanelModal` themselves. Then, the `<router-view />` where those components render can live in the `CreateExamPage/index` component.

Then these components can be shown/hidden depending on whether or not the user is on a "side panel" route or not -- in this case, the only non-side-panel route is the `EXAM_CREATION_ROOT` page component. If we opted to have others, then this display logic can be updated more easily than it is currently.

### Acceptance criteria

- [ ] Remove SectionSidePanel component
- [ ] Handle side panel exiting and/or going back for each page within their relevant component
- [ ] No regressions in quiz creation when navigating side panels

### PR DESCRIPTION
**Pull request by nucleogenesis**
## Summary

https://github.com/user-attachments/assets/47463a48-62f0-48d6-ac04-4d5455315b28

When working on quiz creation I created a SectionSidePanel component to kind of wrap around all of the various side-panel kinds but realized a bit too late on that it wasn't really necessary and added some unnecessary complexity.

This instead begins organizing the code so that each of the child routes' components for the side panels are wrapped in the SidePanelModal component which results in them being rendered into the `<router-view/>` in the quiz creation root route.

Some motivations for this were:

- Removing unnecessary indirection around when/how/where components define what to do re: back and close behavior
- Providing a better example for handling side panels w/ routing for Lessons
- Much easier to put things into the Side panel title section

### QA

- No regressions in the quiz creation side panels - section ordering, section editing, resource selection (search,bookmarks,etc), question replacement.
- Ensure that "are you sure" modals are triggered and that they're "Yes/No" handling works properly.
- Notably - "Section Editor <Click 'Add Questions'>  ->  Resource Selection" - make a change to your selection, then hit the browser back button until you get the "Are you sure" modal -- note that when you confirm it, you are returned correctly to the "Section Editor" side panel. However, had you clicked the "X" to close it, then you'd just go straight to the root page.

### Code review

- Whaddya think overall? 

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue [hunk 1]
```diff
@@ -656,7 +664,7 @@
       }
 
       function handleConfirmClose() {
-        context.emit('closePanel');
+        this.$router.back();
```
Comment (AlexVelezLl) [inline kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue]: I wonder what is the difference between this `this.$router.back()` and the use of the `this.prevRoute`? We could add a comment there to explain the purpose of each and why we have different patterns for confirmClose and just closing the side panel

File: kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue [hunk 2]
```diff
@@ -165,9 +174,10 @@
     components: {
```
Comment (AlexVelezLl) [inline kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue]: I think we can rename these components that renders side panels to be called `ReplaceQuestionsSidePanel` or moving all side panels into a `sidePanels` folder? So externally we can know that all of them are side panels. What do you think?

**Comment by pcenov** [comment]
Hi @nucleogenesis,

Here's a list with the regressions I was able to spot:

- Add questions - The "Are you sure you want to leave" modal is correctly displayed when I click the Close icon after having selected questions, but clicking the Continue button doesn't immediately close the side panel while it should:

https://github.com/user-attachments/assets/098020c3-a2db-4fce-bfe1-0462b1858023

- Add questions - Missing 'Back' arrow after having navigated through a channel's folders. I guess one can still use the browser's back button or the breadcrumb navigation, but it's inconsistent with the implementation at 'Edit section' and not very convenient:

https://github.com/user-attachments/assets/beca343a-d2e9-4d93-9cdd-73cc619cf974

- Add questions - There's not enough spacing between the 'Add questions to 'Section 1' text and the 'Current number of questions in this section'. Long titles will also not be displayed nicely:

space

- Edit section - The "Are you sure you want to leave" modal is displayed when I click the Back button after having selected some questions. In that case it should be possible to go back to the 'Section settings' page:

https://github.com/user-attachments/assets/07ee1bb5-72c5-488d-82af-a9e61a888455

- Edit section > Add more questions - The 'Close' icon is missing so I can't close the side panel right away if I want to:

https://github.com/user-attachments/assets/645cf504-a628-4aca-8b01-7e34e868e903

- Edit section - Changes made to the section title and description are being saved without having clicked the Apply settings button:

https://github.com/user-attachments/assets/ed71c42a-d917-46ca-92bb-591d70f15bf7

- All the modals are not displayed correctly:

https://github.com/user-attachments/assets/94189e55-603d-4c1b-b181-6d9a31f027f1

are u sure

delete

- Selecting a section from the '...' button is not working:

https://github.com/user-attachments/assets/3c0589a9-742e-4e6a-8bf2-24584f166c5c

**Review by AlexVelezLl — COMMENTED** [review]
Just a small consideration regarding the naming of these components, and the "go back" behaviour that is not completely clear to me.

File: kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue (hunk 1 above)
Comment (nucleogenesis) [reply kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ResourceSelection.vue]: The whole thing was kind of confusing so I've included some updates to SidePanelModal that made this a lot easier to deal with. Lemme know what you think when you can.

File: kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue (hunk 2 above)
Comment (nucleogenesis) [reply kolibri/plugins/coach/assets/src/views/plan/CreateExamPage/ReplaceQuestions.vue]: Done, great suggestion, thanks Alex!

**Comment by nucleogenesis** [comment]
@AlexVelezLl @pcenov I have updated the PR and addressed the feedback.

> Edit section - Changes made to the section title and description are being saved without having clicked the Apply settings button:

This is intentional - when you click "Add questions" we decided to skip confirming "Are you sure?" and just auto-save the user's changes when they go to add questions

---

- KModals are all fixed

---

Here is an example of the navigation - I have made changes such that the Back arrow shows AND the Close icon shows in the context of going "Section editor -> Add questions" - so the user can always close out from there but if they use the back arrow, they'll eventually go back to the "Section editor" from the "Add questions" (with the confirmation modal if they've made changes, but no confirmation modal if they have not).

https://github.com/user-attachments/assets/68f42105-33ac-42c1-bc4d-48f42e07bef3


---

I made a change that seems to avoid the "redundant navigation" error. However, I have been able to successfully select any section:

https://github.com/user-attachments/assets/525bdcd8-80a0-4965-891f-2087147e8845

**Comment by pcenov** [comment]
Hi @nucleogenesis - it seems that there are some conflicts that must be resolved and there are no new build assets for me to test at this point, so could you take a look?

**Review by AlexVelezLl — COMMENTED** [review]
Looks good to me! Code changes makes sense to me, and I havent spotted anything weird, I will left the final approve to QA's review. Thanks @nucleogenesis!

**Comment by pcenov** [comment]
Hi @nucleogenesis, looks much better already!

The following issues are still not fixed or partially fixed:

- Add questions - There's not enough spacing between the 'Add questions to 'Section 1' text and the 'Current number of questions in this section'. Long titles will also not be displayed nicely:

https://github.com/user-attachments/assets/45db9b5f-6dd5-4bfb-90a9-bd165544a2a0

- All the modals are now displayed correctly but always with a tiny scroll-bar:

scroll

- Selecting a section from the '...' button is still not working properly as I can't see the first 3 sections when I shrink the window:

https://github.com/user-attachments/assets/40242bd7-2b59-445b-8cec-80b297641a35


> > Edit section - Changes made to the section title and description are being saved without having clicked the Apply settings button:
> 
> This is intentional - when you click "Add questions" we decided to skip confirming "Are you sure?" and just auto-save the user's changes when they go to add questions

I believe that it would be fine to skip the confirmation only if the user has actually added new questions. But if the user has clicked on "Add questions" and then went back to "Section settings" without having added any new questions or clicked the "Apply settings"  button, then it's not ok to auto-save as it's not consistent with the scenario where one goes to "Section settings" makes some changes, hits the close icon and sees the "Are you sure" modal:

https://github.com/user-attachments/assets/e04d98c4-43f2-4388-8f8f-cb6aa66ad99f


> * Notably - "Section Editor <Click 'Add Questions'>  ->  Resource Selection" - make a change to your selection, then hit the browser back button until you get the "Are you sure" modal -- note that when you confirm it, you are returned correctly to the "Section Editor" side panel. However, had you clicked the "X" to close it, then you'd just go straight to the root page.


When following your guidance for this scenario I am always going back to the root page, and not to the "Section settings" side panel:

https://github.com/user-attachments/assets/90040059-6abb-4194-9f26-77433893aa58

File: kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue [hunk 3]
```diff
@@ -491,9 +485,6 @@
         loadingMore,
       } = useQuizResources({ topicId, practiceQuiz: selectPracticeQuiz.value });
 
-      const { searchTerms, search } = useBaseSearch({ descendant: topic });
```
Comment (AlexVelezLl) [inline kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue:494]: This wasnt being used before?

File: kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue [hunk 4]
```diff
@@ -823,20 +832,34 @@
...
       },
       /*
-    selectAllIsVisible() {
-      TO BE IMPLEMENTED IN https://github.com/learningequality/kolibri/issues/11734
-      Should only be visible if there are any checkboxes at all -- we'll only be showing
-      checkboxes for Exercises, not topics
-    },
-    */
+        selectAllIsVisible() {
+          TO BE IMPLEMENTED IN https://github.com/learningequality/kolibri/issues/11734
```
Comment (AlexVelezLl) [inline kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue:836]: This issue is already closed. Its still valid? Or do we need to remove this?

File: kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue [hunk 5]
```diff
@@ -712,8 +704,26 @@
...
+      const canGoBack = computed(() => {
+        return Boolean(
+          // If we can go back to the section editor (another side panel) we can go back from here
+          // The beforeRouteLeave will catch and ask the user to confirm that navigation if needed
+          (prevRoute.value.name && prevRoute.value.name === PageNames.QUIZ_SECTION_EDITOR) ||
+            // prevRoute.value.name is falsy if we only just loaded, so we'll never show the back
+            // arrow in that case. Here we check if there is a valid prevRoute and if there is,
+            // then we know we can go back if we have a topicId or are showing bookmarks
+            // TODO When we add search it'll probably want to be here too
+            (prevRoute.value.name && (topicId.value || showBookmarks.value)),
```
Comment (AlexVelezLl) [inline kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue:720]: I think its a very edge case, but with this condition, if I reload the page, `prevRoute.value.name` is always empty, and I completely lose the back button and doesnt appear again even if I navigate through the tree:

https://github.com/user-attachments/assets/7d56749e-6cec-4461-a64d-b58390b86c41

File: kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue [hunk 6]
```diff
@@ -848,6 +871,12 @@
       focusFirstEl() {
         this.$refs.textbox.focus();
       },
+      handleGoBack() {
+        this.$router.back();
```
Comment (AlexVelezLl) [inline kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue:875]: I was wondering what does the "go back" button means when we are navigating through the tree. If it means the same behaviour as the browser back button, or if it means go back in the tree. From my perspective I would think that if I click the back button while searching through the tree, it always take me back to the parent folder, but this doesnt happen if we click on the breadcrumbs:


https://github.com/user-attachments/assets/918da58b-91af-47f8-891b-cd7ad2a3f1d7

Comment (AlexVelezLl) [reply kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue:875]: It also happens when we open the section editor:

https://github.com/user-attachments/assets/1db4fd4a-40ce-4f5d-b1a7-3664abe06a88

**Review by AlexVelezLl — COMMENTED** [review]
Thanks Jacob! I have spotted a couple of weird behaviours with the back button, but they are mostly questions to confirm if this is the expected behaviour

File: kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue (hunk 5 above)
Comment (nucleogenesis) [reply kolibri/plugins/coach/assets/src/views/quizzes/CreateExamPage/SidePanels/ResourceSelection.vue:720]: Thank you for this - it shouldn't show up on page load, but it should once you start navigating for sure

**Comment by nucleogenesis** [comment]
Despite the time I've put into this I think that ultimately the myriad issues that have arisen w/ regard to going back/forward and which options to show and how/where we show the heading section and so forth --- many of the changes here are going to be obsolete.

The simplest possible change here is to just move the SidePanel components into their own folder, but ultimately the major fix I wanted here was to clean up the routing/navigation business. 

So, given the upcoming refactor of side panels I think that this can be closed for now and we can take a fresh look at this whole issue next week again in collaboration.

**Pull request closed without merging**